onnxruntime/core/xnnpack/optimizer/xnnpack_transformer.cc

```diff
@@ -0,0 +1,497 @@
+                         const std::string& name, const T& value) {
+  attrs[name] = utils::MakeAttribute(name, value);
+}
+
+static Status ReplaceMaxPool(Graph& main_graph, Node& nodeRef, bool& modified) {
+  ProtoHelperNodeContext nc(nodeRef);
+  OpNodeProtoHelper info(&nc);
+  PoolAttributes pool_attrs(info, "MaxPool", nodeRef.SinceVersion());
+
+  auto inputdefs = nodeRef.InputDefs();
+  if (inputdefs.size() != 1) return Status::OK();
+
+  // Skip if unknown rank
+  auto shape = inputdefs[0]->Shape();
+  if (shape == nullptr || shape->dim_size() != 4) {
+    return Status::OK();
+  }
+
+  modified = true;
+
+  std::unordered_map<std::string, ONNX_NAMESPACE::AttributeProto> attrs;
+  AddAttribute(attrs, "input_padding_top", pool_attrs.pads[0]);
+  AddAttribute(attrs, "input_padding_right", pool_attrs.pads[3]);
+  AddAttribute(attrs, "input_padding_bottom", pool_attrs.pads[2]);
+  AddAttribute(attrs, "input_padding_left", pool_attrs.pads[1]);
+
+  AddAttribute(attrs, "pooling_height", pool_attrs.kernel_shape[0]);
+  AddAttribute(attrs, "pooling_width", pool_attrs.kernel_shape[1]);
+
+  AddAttribute(attrs, "stride_height", pool_attrs.strides[0]);
+  AddAttribute(attrs, "stride_width", pool_attrs.strides[1]);
+
+  AddAttribute(attrs, "dilation_height", pool_attrs.dilations[0]);
+  AddAttribute(attrs, "dilation_width", pool_attrs.dilations[1]);
+  if (pool_attrs.auto_pad == AutoPadType::SAME_UPPER) {
+    AddAttribute(attrs, "padding_mode", static_cast<int64_t>(1));
+  } else {
+    AddAttribute(attrs, "padding_mode", static_cast<int64_t>(0));
+  }
+
+  Node* new_node = nullptr;
+  XNNPACK_RETURN_IF_ERROR(
+      ReplaceNode(main_graph, nodeRef, "XnnPackMaxPooling2d", "", &attrs, "com.microsoft", &new_node));
+  XNNPACK_RETURN_IF_ERROR(TranposeNCHWToNHWC(main_graph, 4, *new_node));
+  return Status::OK();
+}
+
+static Status ReplaceConv(Graph& main_graph, Node& nodeRef, bool& modified) {
+  // The next node after ONNX Conv
+  const Node* const next_node_to_fuse =
+      optimizer_utils::CheckOutputEdges(main_graph, nodeRef, 1) ? &*nodeRef.OutputNodesBegin() : nullptr;
+
+  ProtoHelperNodeContext nc(nodeRef);
+  OpNodeProtoHelper info(&nc);
+  int64_t group = 1;
+  XNNPACK_RETURN_IF_ERROR(info.GetAttr<int64_t>("group", &group));
+  auto X_input = info.GetInputType(0);
+  auto weight_input = info.GetInputType(1);
+  TensorShape weight_shape = utils::GetTensorShapeFromTensorShapeProto(weight_input->tensor_type().shape());
+  TensorShape X_shape = utils::GetTensorShapeFromTensorShapeProto(X_input->tensor_type().shape());
+  if (weight_shape.NumDimensions() != 4 || X_shape.NumDimensions() != 4) return Status::OK();
+
+  // Now the input shape is still NCHW
+  int64_t input_channels = X_shape[1];
+
+  if (group != 1 && group != input_channels) return Status::OK();
+  for (size_t i = 0; i != weight_shape.NumDimensions(); ++i) {
+    if (weight_shape[i] <= 0) return Status::OK();
+  }
```

Comment:
Can we move all those checks to IsConvSupportedByXNNPack and we don't need to check again here

Reply:
The graph may have changed. I can call the same function again here. But the check is still needed.

Reply:
Why do we need to check again? Supposedly, those should still be valid if IsConvSupportedByXNNPack already check it? ShapeInference results are broken somewhere?

Reply:
> ShapeInference results are broken somewhere?

Roughly. It's possible that a dim of a tensor was known before this transformer making changes but it becomes unknown after. We can't really say the ShapeInference function is broken, because shape inference functions just do their best effort. As long as they don't generate wrong numbers, they are not broken.